### ISSUE TITLE
Dispatching to a decoder based on Content-Type header

### ISSUE THREAD

**Issue by vkostyukov** [finagle/finch#958]
This came out of [this Gitter discussion](https://gitter.im/finagle/finch?at=5b3e7ce0bd92d80782971b48).

As of now, Finch's `body*` endpoints don't really care about input's content-type as it's only being used to lookup decoders at compile time. It would be awesome if we can employ the fact we can resolve multiple decoders at compile time, and dispatch input to them later, when endpoint is being run.

Something along the lines:

```scala
body[Application.Json :+: Text.Plain :+: CNil] <- dispatches to a JSON or text decoder based on the input's content type
```

See #905 for an inspiration.

**Comment by vkostyukov** [comment]
That's good with me, @sergeykolbasov!

I want to aim to ship this in the next release. It /seems/ possible given we already did a very similar trick on the outbound path.

**Comment by vkostyukov** [comment]
Implemented in #966.
